superset-frontend/src/components/DynamicPlugins/DynamicPluginProvider.tsx

```diff
@@ -0,0 +1,181 @@
+  // Maybe someone else can help figure that out.
+  const loadingKey = `__superset__loading__/${name}`;
+  const packageKey = `__superset__/${name}`;
+  if (window[loadingKey]) {
+    await window[loadingKey];
+    return window[packageKey];
+  }
+  window[loadingKey] = promise;
+  const pkg = await promise;
+  window[packageKey] = pkg;
+  return pkg;
+}
+
+async function defineSharedModules(moduleMap: { [key: string]: Promise<any> }) {
+  return Promise.all(
+    Object.entries(moduleMap).map(([name, promise]) => {
+      defineSharedModule(name, promise);
+      return promise;
+    }),
+  );
+}
+
+type CompleteAction = {
+  type: 'complete';
+  key: string;
+  error: null | Error;
+};
+
+type BeginAction = {
+  type: 'begin';
+  keys: string[];
+};
+
+function pluginContextReducer(
+  state: PluginContextType,
+  action: BeginAction | CompleteAction,
+): PluginContextType {
+  switch (action.type) {
+    case 'begin': {
+      const plugins = { ...state.plugins };
+      for (const key of action.keys) {
+        plugins[key] = { key, error: null, loading: true };
+      }
+      return {
+        ...state,
+        loading: true,
+        plugins,
+      };
+    }
+    case 'complete': {
+      return {
+        ...state,
+        loading: Object.values(state.plugins).some(
+          plugin => plugin.loading && plugin.key !== action.key,
+        ),
+        plugins: {
+          ...state.plugins,
+          [action.key]: {
+            key: action.key,
+            loading: false,
+            error: action.error,
+          },
+        },
+      };
+    }
+    default:
+      return state;
+  }
+}
+
+export type Props = React.PropsWithChildren<{}>;
+
+export default function DynamicPluginProvider({ children }: Props) {
+  const [pluginState, dispatch] = useReducer(pluginContextReducer, {
+    // use the dummy plugin context, and override the methods
+    ...dummyPluginContext,
+    // eslint-disable-next-line @typescript-eslint/no-use-before-define
+    fetchAll,
+    loading: isFeatureEnabled(FeatureFlag.DYNAMIC_PLUGINS),
+    // TODO: Write fetchByKeys
+  });
+
+  async function fetchAll() {
+    try {
+      await defineSharedModules({
+        react: import('react'),
+        lodash: import('lodash'),
+        'react-dom': import('react-dom'),
+        '@superset-ui/chart': import('@superset-ui/chart'),
+        '@superset-ui/chart-controls': import('@superset-ui/chart-controls'),
+        '@superset-ui/connection': import('@superset-ui/connection'),
+        '@superset-ui/color': import('@superset-ui/color'),
+        '@superset-ui/core': import('@superset-ui/core'),
+        '@superset-ui/dimension': import('@superset-ui/dimension'),
+        '@superset-ui/query': import('@superset-ui/query'),
+        '@superset-ui/style': import('@superset-ui/style'),
+        '@superset-ui/translation': import('@superset-ui/translation'),
+        '@superset-ui/validator': import('@superset-ui/validator'),
+      });
```

Comment:
Should `defineSharedModules` be called outside of `fetchAll` or in a separate `setup` function? It would also be nice to make `DynamicProviderProvider` load only specific plugins on demand instead of all together.

Reply:
I would like to do loading on demand, but the way I got the example plugin's webpack to work is it references global variables on the `window`. I intend to define a `superset-plugin.json` manifest file for dynamic plugins in the future, which would be a good place to put dependency info. For now I'd like to keep this minimal.

Reply:
This design is also not ideal for third-party plugins that may want to add libraries for other plugins to depend on. Definitely should be improved.